### PR TITLE
add kafka producer

### LINKED ISSUE #18: #### Kafka producer
Create a component for producing kafka messages to a configured topic
Producer has to be sychronous.
Produced messages should have ability to be consumed in order from other side.

### PR DESCRIPTION
**Pull request by crnopster**

### CODE REVIEW

File: kafka/producer.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	return &producer{
+		KafkaProducer: saramaProducer,
+		Config:        config,
+	}, nil
+}
+
+// Produce message with filtered position to kafka topic "filtered-positions"
+func (p *producer) Produce(pos position.Position) error {
```
Comment (dmitriyminer) [inline kafka/producer.go:41]: It would be great to declare Message type

Comment (crnopster) [reply kafka/producer.go:41]: fixed

File: kafka/producer_test.go [hunk 2]
```diff
@@ -0,0 +1,118 @@
...
+	}
+	got, err := NewProducer(config)
+
+	if err != nil {
+		t.Errorf("NewProducer() got %v, want nil", err)
+	}
+
+	if got == nil {
+		t.Errorf("NewProducer() got nil, want Producer")
+	}
```
Comment (radikh) [inline kafka/producer_test.go]: Use some assertion library for checks.

File: kafka/producer_test.go [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+	got, err := NewProducer(config)
+	if err == nil {
+		t.Errorf("NewProducer() got nil, want error")
+		return
+	}
+
+	if got != nil {
+		t.Errorf("NewProducer() got Producer, want nil")
+	}
```
Comment (radikh) [inline kafka/producer_test.go]: Use some assertion library for checks.

File: kafka/producer_test.go [hunk 4]
```diff
@@ -0,0 +1,118 @@
...
+		KafkaProducer: kafkaProducer,
+		Config:        config,
+	}
+
+	kafkaProducer.ExpectSendMessageAndSucceed()
+
+	var p position.Position
+	if err := producer.Produce(p); err != nil {
+		t.Errorf("Producer.Produce() got %v, want nil", err)
+	}
```
Comment (radikh) [inline kafka/producer_test.go]: Use some assertion library for checks.

File: kafka/producer_test.go [hunk 5]
```diff
@@ -0,0 +1,118 @@
...
+
+	lat := float32(math.Inf(1))
+	p := position.Position{
+		UserID:   uuid.New(),
+		Latitude: lat,
+	}
+
+	if err := producer.Produce(p); err == nil {
+		t.Errorf("Producer.Produce() got nil, want %v", testError)
+	}
```
Comment (radikh) [inline kafka/producer_test.go]: Use some assertion library for checks.

File: config/config_test.go [hunk 6]
```diff
@@ -95,3 +96,22 @@ func TestConfigNewLoggerConfig(t *testing.T) {
...
+		Host:            "HOST4",
+		Port:            "PORT4",
+		Version:         "V4",
+		ConsumerTopic:   "ConsumerTopic",
+		ConsumerGroupID: "ConsumerGroupID",
+		ProducerTopic:   "ProducerTopic",
+	}
+	if got := v.NewKafkaConfig(); !reflect.DeepEqual(got, want) {
+		t.Errorf("Config.NewKafkaConfig() = %v, want %v", got, want)
+	}
```
Comment (radikh) [inline config/config_test.go]: Use some assertion library for checks.

File: config/config_test.go [hunk 7]
```diff
@@ -95,3 +96,22 @@ func TestConfigNewLoggerConfig(t *testing.T) {
 		t.Errorf("Config.NewLoggerConfig() = %v, want %v", got, want)
 	}
 }
+
+func TestConfigNewKafkaConfig(t *testing.T) {
+	v, err := NewConfig("testConfigForViper", "./testData/")
+	if err != nil {
+		t.Fatalf("Cant start test, err: %v", err)
+	}
```
Comment (radikh) [inline config/config_test.go]: Use some assertion library for checks.

File: kafka/config.go [hunk 8]
```diff
@@ -0,0 +1,11 @@
+// Package kafka provides producer and consumer to work with kafka topics
+package kafka
+
+type Config struct {
```
Comment (radikh) [inline kafka/config.go:5]: Config should have a comment.

File: kafka/producer.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+
+import (
+	"encoding/json"
+	"fmt"
+
+	"github.com/Shopify/sarama"
+	"github.com/lvl484/positioning-filter/position"
+)
+
+type Producer interface {
```
Comment (radikh) [inline kafka/producer.go:15]: Producer should have a comment.

File: kafka/producer.go [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+	Produce(position.Position) error
+	Close() error
+}
+
+type producer struct {
+	KafkaProducer sarama.SyncProducer
+	Config        *Config
+}
+
+func NewProducer(config *Config) (Producer, error) {
```
Comment (radikh) [inline kafka/producer.go:26]: In this case, that is not so important, but when some pointers are passed to a func you have to be sure that they are concurrently safe.

File: kafka/producer.go [hunk 11]
```diff
@@ -0,0 +1,59 @@
...
+	}
+
+	if _, _, err := p.KafkaProducer.SendMessage(msg); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *producer) Close() error {
```
Comment (radikh) [inline kafka/producer.go:62]: Close should have a comment.

File: kafka/producer_test.go [hunk 12]
```diff
@@ -0,0 +1,118 @@
...
+	config := &Config{
+		Host: "somehost",
+		Port: "1",
+	}
+
+	got, err := NewProducer(config)
+	if err == nil {
+		t.Errorf("NewProducer() got nil, want error")
+		return
+	}
```
Comment (radikh) [inline kafka/producer_test.go]: Use some assertion library for checks.

File: kafka/producer_test.go [hunk 13]
```diff
@@ -0,0 +1,118 @@
...
+}
+
+func TestProducerClose(t *testing.T) {
+	kafkaProducer := saramaMocks.NewSyncProducer(t, nil)
+	producer := producer{
+		KafkaProducer: kafkaProducer,
+	}
+
+	if err := producer.Close(); err != nil {
+		t.Errorf("Close() got %v, want nil", err)
```
Comment (radikh) [inline kafka/producer_test.go]: Use some assertion library for checks.

File: kafka/producer_test.go [hunk 14]
```diff
@@ -0,0 +1,118 @@
...
+	}
+
+	testError := errors.New("Test error for Producer")
+
+	kafkaProducer.ExpectSendMessageAndFail(testError)
+
+	var p position.Position
+	if err := producer.Produce(p); err == nil {
+		t.Errorf("Producer.Produce() got nil, want %v", testError)
+	}
```
Comment (radikh) [inline kafka/producer_test.go]: Use some assertion library for checks.

File: kafka/producer_test.go [hunk 15]
```diff
@@ -0,0 +1,118 @@
...
+	kafkaProducer := saramaMocks.NewSyncProducer(t, nil)
+	config := &Config{
+		ProducerTopic: "Topic",
+	}
+	producer := producer{
+		KafkaProducer: kafkaProducer,
+		Config:        config,
+	}
+
+	kafkaProducer.ExpectSendMessageAndSucceed()
```
Comment (radikh) [inline kafka/producer_test.go]: Message has to be  checked if it has expected value;

File: kafka/producer_test.go [hunk 16]
```diff
@@ -0,0 +1,118 @@
...
+		ProducerTopic: "Topic",
+	}
+	producer := producer{
+		KafkaProducer: kafkaProducer,
+		Config:        config,
+	}
+
+	testError := errors.New("Test error for Producer")
+
+	kafkaProducer.ExpectSendMessageAndFail(testError)
```
Comment (radikh) [inline kafka/producer_test.go]: Message has to be  checked if it has expected value;

File: kafka/producer_test.go (hunk 2 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 3 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 4 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 5 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 12 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 13 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 14 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 15 above)
Comment (crnopster) [reply kafka/producer_test.go]: Message can't be checked in this case

File: kafka/producer_test.go (hunk 16 above)
Comment (crnopster) [reply kafka/producer_test.go]: Message can't be checked in this case

File: config/config_test.go (hunk 6 above)
Comment (crnopster) [reply config/config_test.go]: fixed

File: config/config_test.go (hunk 7 above)
Comment (crnopster) [reply config/config_test.go]: fixed

File: kafka/config.go (hunk 8 above)
Comment (crnopster) [reply kafka/config.go:5]: fixed

File: kafka/producer.go (hunk 9 above)
Comment (crnopster) [reply kafka/producer.go:15]: fixed

File: kafka/producer.go (hunk 11 above)
Comment (crnopster) [reply kafka/producer.go:62]: fixed

**Review by alpaka-alpachino — DISMISSED** [review]

**Review by maksymbilokur — DISMISSED** [review]

File: kafka/producer_test.go (hunk 15 above)
Comment (radikh) [reply kafka/producer_test.go]: You can use this call: https://godoc.org/github.com/Shopify/sarama/mocks#SyncProducer.ExpectSendMessageWithCheckerFunctionAndSucceed

File: kafka/producer_test.go (hunk 16 above)
Comment (radikh) [reply kafka/producer_test.go]: You can use this call: https://godoc.org/github.com/Shopify/sarama/mocks#SyncProducer.ExpectSendMessageWithCheckerFunctionAndFail

File: kafka/producer_test.go (hunk 15 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 16 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

**Review by alpaka-alpachino — DISMISSED** [review]

**Review by maksymbilokur — DISMISSED** [review]

File: kafka/producer_test.go [hunk 17]
```diff
@@ -0,0 +1,122 @@
...
+		ProducerTopic: "Topic",
+	}
+	producer := producer{
+		KafkaProducer: kafkaProducer,
+		Config:        config,
+	}
+
+	testError := errors.New("Test error for Producer")
+
+	var p position.Position
```
Comment (radikh) [inline kafka/producer_test.go]: So position is empty, how do you know that functions work well and don't always expose empty position?

File: kafka/producer_test.go [hunk 18]
```diff
@@ -0,0 +1,122 @@
...
+	kafkaProducer := saramaMocks.NewSyncProducer(t, nil)
+	config := &Config{
+		ProducerTopic: "Topic",
+	}
+	producer := producer{
+		KafkaProducer: kafkaProducer,
+		Config:        config,
+	}
+
+	var p position.Position
```
Comment (radikh) [inline kafka/producer_test.go]: So position is empty, how do you know that functions work well and don't always expose empty position?

File: kafka/producer_test.go (hunk 17 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

File: kafka/producer_test.go (hunk 18 above)
Comment (crnopster) [reply kafka/producer_test.go]: fixed

**Review by radikh — APPROVED** [review]

**Review by alpaka-alpachino — APPROVED** [review]

**Review by oklymchuk — APPROVED** [review]

**Review by maksymbilokur — APPROVED** [review]

**Pull request merged**